### PR TITLE
feat(optimize): score keyword candidates against official Apple sources

### PR DESCRIPTION
**Pull request by rudrankriyam**
> **Stacked PR.** Base is `feat/optimize-keywords-rank` (#2120). Review only the last commit; this retargets to `main` automatically when #2120 merges.

## Problem

`asc optimize keywords rank` (#2120) answers "where do I rank for this keyword". It does not answer the question that decides whether a keyword is worth pursuing at all: *how hard is this keyword, and does anyone search it.* Both answers exist across surfaces this repository already ships — public App Store search, public app metadata, and the official Apple Ads demand endpoint — but nothing composes them per keyword.

## Behavior

```
asc optimize keywords score --keywords LIST [--country us] [--app APP_ID] [--genre GENRE] [--ad-account ID] [--ads-profile NAME] [--workers N]
```

Each keyword is evaluated against three independent sources, each reporting its own status with the same `available` / `empty` / `unavailable` vocabulary the Apple Ads optimization surface already uses:

| Source | Requires | Contributes |
| --- | --- | --- |
| `public_search` | nothing | competitor ordering, `appCount` |
| `competitor_metadata` | nothing | competitor release dates |
| `search_term_popularity` | `--genre`, Ads credentials | `popularity` |
| `app_rank` | `--app` | this app's position |

Competition alone is enough to produce a difficulty score, so the command is useful with zero authentication.

**Degrade, don't fabricate.** Without `--genre` and Ads credentials, `popularity` is `null` and the source status names the missing flags — everything else still computes. If the metadata lookup fails, the two date-derived signals fall back to a documented one-year window, the source is marked unavailable, and the affected `rawSignals` entries carry empty date strings so the degradation is visible rather than implied. A keyword whose own search fails becomes an `unavailable` row whose `difficultyScore`, `minDifficultyScore`, `isBrandKeyword`, and `appCount` serialize as explicit `null` rather than as zeros. The command fails only when every keyword fails.

Searches fan out with the `--workers` bound from #2120. Competitor metadata is deduplicated across every keyword and fetched in batches of at most 50, so 100 keywords cost ~10 lookup requests rather than 100.

### Example

```console
$ asc optimize keywords score --keywords "focus timer" --app 1234567890
```

```json
{
  "schemaVersion": "1", "appId": "1234567890", "country": "US",
  "sources": [
    { "name": "app_rank", "status": "available", "count": 1 },
    { "name": "competitor_metadata", "status": "available", "count": 5 },
    { "name": "public_search", "status": "available", "count": 1 },
    { "name": "search_term_popularity", "status": "unavailable",
      "error": "Apple Ads search popularity needs --genre; popularity was not requested" }
  ],
  "rows": [{
    "keyword": "focus timer", "status": "available",
    "popularity": null,
    "difficultyScore": 47.5, "minDifficultyScore": 40.0,
    "isBrandKeyword": false, "appCount": 120, "keywordMatch": "titleExactPhrase", "rank": 5,
    "averageAppScore": 0.6, "minimumAppScore": 0.4, "normalizedAppCount": 0.5789,
    "rawSignals": [{
      "appId": "111", "name": "Focus Timer", "subtitle": "", "publisherName": "Alpha Labs",
      "averageUserRating": 4.6, "userRatingCount": 8000,
      "releaseDate": "2020-01-15T08:00:00Z", "currentVersionReleaseDate": "2026-08-01T08:00:00Z",
      "daysSinceFirstRelease": 2408, "daysSinceLastRelease": 18,
      "normalizedRatingCount": 0.8, "normalizedAverageRating": 0.8, "normalizedAge": 0.9507,
      "ratingsPerDay": 3.32, "normalizedRatingsPerDay": 0.2676,
      "keywordMatch": "titleExactPhrase", "keywordMatchScore": 1, "appScore": 0.6259
    }]
  }]
}
```

`--output table` and `--output markdown` render summary, keyword, and source sections.

## Formula transparency

The anti-invention rule is the point of this PR, not a footnote. Every computed score ships alongside the named raw inputs it was derived from, under `rows[].rawSignals[]` — every normalized signal, its unnormalized source value, and the per-app score. A reader can re-derive any difficulty by hand from one JSON payload without re-running the command.

The full formula, every constant, the match ladder, the brand heuristic, and the limitations are documented in **`docs/design/optimize-keywords.md`**. Two worked parity vectors from that document are pinned as tests, so the formula cannot drift silently:

- App `{avg 4.5, count 1000, last release 30d, first release 400d, titleAllWords}` → `nRatingCount` 0.1, `nAvgRating` 0.75, `nAge` 0.9178, `nRPD` 0.26136, `appScore` 0.50519
- Keyword `{appScores [.8,.7,.6,.5,.4], appCount 120}` → `avg` 0.6, `min` 0.4, `nAppCount` 0.57895, `difficulty` 47.5304, `minDifficulty` 40

Both match to within 1e-9.

### Attribution

The scoring methodology — the difficulty formula and its constants, the keyword match ladder, and the brand heuristic — is **adapted from semihcihan's App Store Optimization CLI**, which is MIT licensed:

<https://github.com/semihcihan/App-Store-Optimization-CLI>

`internal/cli/optimize/keywords_difficulty.go` is an independent Go implementation of that published formula, not ported code. The credit appears in the design document, in a header comment on the engine source file, and in the README acknowledgements. The repository has no `NOTICE` or `THIRD_PARTY` file convention, so the existing "Acknowledgements" section in `README.md` was extended in the same format as the entries already there.

**Heads-up on `README.md`:** open PR #2062 also touches that file. My hunk is a 3-line addition at the end of the existing Acknowledgements section near the bottom, so any conflict should be trivial.

## Documented limitations

Stated in the design doc rather than papered over:

- **Subtitle is not available from the public endpoints.** Apple's public lookup response has no subtitle field, so matches resolve on the title alone and the subtitle rungs of the ladder stay unreachable through this source. `rawSignals[].subtitle` is emitted as an empty string so the gap is visible. The ladder still implements and tests those rungs, so a future subtitle source drops straight in.
- **Popularity has no storefront dimension of its own** and is strongly US-centric; it is reported with the country, genre, and publication week it came from, never rescaled to other storefronts.
- **Public endpoints are volatile** — a score is a snapshot of one observation, which is why `generatedAt` and every raw input travel with it.
- **`appCount` saturates at 200**, matching the public search request cap.

## Adaptations worth a maintainer's attention

Two premises shifted during implementation. Both are documented in the design doc's limitations section as natural follow-ups:

1. **`--genre` was added.** Popularity is scoped by country and genre, so bare keyword lists can request demand without `--app`. The app flag is reserved for rank for the selected app. When genre or credentials are absent, the source reports exactly what is missing.
2. **Release dates are read directly.** `internal/itunes.App` does not expose `releaseDate` or `currentVersionReleaseDate`, and I deliberately did not widen that shared type while other work is in flight there. `internal/cli/optimize/keywords_metadata.go` reads those two fields from the same public lookup endpoint using that client's own base URL and HTTP client — an isolated ~60-line seam. Folding it into `internal/itunes` once the in-flight work lands is the obvious cleanup.

## Tests

- `internal/cli/optimize/keywords_difficulty_test.go` — both mandatory parity vectors, per-signal normalization edges (clamps, rating-count damping, stale releases, missing and unparseable dates, day flooring, velocity ceiling), difficulty fallback/clamp/saturation, the full match ladder in order including NFKC compatibility folding and accent preservation, and the brand heuristic including the exact median boundary.
- `internal/cli/optimize/keywords_score_test.go` — `httptest` coverage for full compose, thin-window fallback, per-keyword failure isolation, Ads popularity flattening (most recent week wins), ads-unavailable degrade, metadata-unavailable degrade, all-failed error, registered table and markdown rendering without a JSON fallback, and every usage error asserted to run before any client is built.
- `internal/cli/cmdtest/optimize_keywords_score_test.go` — end-to-end through the real root command: help wiring, usage errors, the full JSON contract with raw-signal reproducibility, deduplicated lookup batching, explicit-null serialization, fallback, and brand detection.

Gauntlet: `make build`, `make format`, `make check-docs`, `make lint` (0 issues), `ASC_BYPASS_KEYCHAIN=1 make test` — all green. No live API calls; all coverage is `httptest`/round-tripper fixtures.

## Compatibility

Additive only. New leaf under the existing experimental `optimize keywords` group; no existing command, flag, or output shape changes. `docs/COMMANDS.md` lists top-level commands only, so it is unchanged (verified). As in #2120, the `[experimental]` marker uses this tree's trailing-suffix convention rather than the prefix convention asserted by `stability_tiers_test.go` for the surfaces it enumerates — matched to the siblings deliberately, and happy to flip it tree-wide if you prefer.

## Summary by CodeRabbit

- **New Features**
  - Added the experimental, read-only `asc optimize keywords score` command.
  - Evaluates keyword difficulty, competition, brand relevance, rankings, and optional popularity.
  - Supports country, app, genre, and worker configuration.
  - Provides table, Markdown, and JSON reports with source availability and partial-result indicators.
  - Handles sparse or unavailable data with documented fallback behavior.
  - Apple Ads popularity can be collected without specifying an app.

- **Documentation**
  - Added design documentation covering scoring, data sources, limitations, and output behavior.
  - Added attribution for the keyword-difficulty methodology.

### CODE REVIEW

File: internal/cli/optimize/keywords_difficulty.go [hunk 1]
```diff
@@ -0,0 +1,385 @@
...
+// caller can re-derive the score without re-running the command.
+type KeywordScoreSignals struct {
+	AppID                     string  `json:"appId"`
+	Name                      string  `json:"name"`
+	Subtitle                  string  `json:"subtitle"`
+	PublisherName             string  `json:"publisherName"`
+	AverageUserRating         float64 `json:"averageUserRating"`
+	UserRatingCount           int64   `json:"userRatingCount"`
+	ReleaseDate               string  `json:"releaseDate,omitempty"`
+	CurrentVersionReleaseDate string  `json:"currentVersionReleaseDate,omitempty"`
```
[the remark this answers is not in the document]
Comment (rudrankriyam) [reply internal/cli/optimize/keywords_difficulty.go]: Fixed in eeedbcdca. The centralized raw-signal contract no longer omits either release-date field, and a regression test marshals degraded signals and asserts that both keys remain present with empty-string values. The full local gate set passes.

File: internal/cli/optimize/keywords_score.go [hunk 2]
```diff
@@ -0,0 +1,611 @@
...
+	Keywords     int `json:"keywords"`
+	Scored       int `json:"scored"`
+	Unavailable  int `json:"unavailable"`
+	WithRank     int `json:"withRank"`
+	BrandMatches int `json:"brandMatches"`
+}
+
+// KeywordScoreReport is the stable JSON contract emitted by
+// `asc optimize keywords score`.
+type KeywordScoreReport struct {
```
[the remark this answers is not in the document]
Comment (rudrankriyam) [reply internal/cli/optimize/keywords_score.go]: Fixed in eeedbcdca. The score report and its nested output types now live in internal/asc/output_keyword_score.go, its multi-table renderer is registered centrally, and the command uses shared.PrintOutput. A focused regression test now fails if table or Markdown rendering falls back to JSON. The full local gate set passes.

**Comment by rudrankriyam** [comment]
@codex review

**Comment by rudrankriyam** [comment]
@coderabbitai review

File: internal/cli/optimize/keywords_difficulty.go [hunk 3]
```diff
@@ -0,0 +1,362 @@
...
+	}
+	return strings.Join(strings.Fields(builder.String()), " ")
+}
+
+// computeKeywordDifficulty aggregates competitor app scores into a keyword
+// difficulty. The minimum app score is weighted most heavily because the
+// weakest app in the top results is the realistic entry point.
+func computeKeywordDifficulty(appScores []float64, appCount int) keywordDifficultyResult {
+	if appCount < keywordFallbackApps {
+		return keywordDifficultyResult{Difficulty: 1, MinDifficulty: 1, Fallback: true}
```
[the remark this answers is not in the document]
Comment (rudrankriyam) [reply internal/cli/optimize/keywords_difficulty.go]: Fixed in 9c9589c23. Fallback still pins difficulty and minimumDifficulty to 1 but now preserves the observed average, minimum, and normalized app count. Unit and root-command JSON tests cover this edge case.

File: internal/cli/cmdtest/optimize_keywords_score_test.go [hunk 4]
```diff
@@ -0,0 +1,357 @@
...
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			t.Setenv("ASC_BYPASS_KEYCHAIN", "1")
+			t.Setenv("ASC_APP_ID", "")
+			originalTransport := http.DefaultTransport
+			t.Cleanup(func() { http.DefaultTransport = originalTransport })
+			http.DefaultTransport = roundTripFunc(func(req *http.Request) (*http.Response, error) {
+				t.Fatalf("unexpected request before input validation: %s", req.URL.String())
+				return nil, errors.New("unexpected request")
+			})
```
[the remark this answers is not in the document]
Comment (rudrankriyam) [reply internal/cli/cmdtest/optimize_keywords_score_test.go:131]: Fixed in 9c9589c23. RoundTripper callbacks now use Errorf and return immediately instead of Fatalf from worker goroutines; focused and full tests pass.

File: docs/design/optimize-keywords.md [hunk 5]
```diff
@@ -0,0 +1,262 @@
...
+nRatingCount = clamp(userRatingCount / 10000, 0, 1)
+nAvgRating   = avgRating <= 3 ? 0
+             : clamp((avgRating - 3) / 2, 0, 1) * min(userRatingCount, 20) / 20
+nAge         = 1 - clamp(daysSinceLastRelease / 365, 0, 1)
+rpd          = userRatingCount / daysSinceFirstRelease
+nRPD         = rpd <= 0   ? 0
+             : rpd <= 1   ? rpd * 0.25
+             : rpd < 100  ? 0.25 + 0.75 * ((rpd - 1) / 99)
+             :              1
+```
```
[the remark this answers is not in the document]
Comment (rudrankriyam) [reply docs/design/optimize-keywords.md]: Fixed in 9c9589c23. All three formula fences now use text language annotations; documentation checks pass.

File: internal/cli/optimize/keywords_score.go [hunk 6]
```diff
@@ -0,0 +1,559 @@
...
+// demand snapshot. The source is structurally scoped to an app and a genre, so
+// it degrades to unavailable with a reason whenever either is missing.
+func collectKeywordPopularity(
+	ctx context.Context,
+	request keywordPopularityRequest,
+) (map[string]asc.KeywordPopularity, ads.SearchOptimizationSourceStatus) {
+	missing := make([]string, 0, 2)
+	if strings.TrimSpace(request.AppID) == "" {
+		missing = append(missing, "--app")
+	}
```
[the remark this answers is not in the document]
Comment (rudrankriyam) [reply internal/cli/optimize/keywords_score.go]: Fixed in 9c9589c23. Added a popularity-only authenticated collection path; keyword scoring now requires only --genre for demand and --app only for rank. Regression tests verify empty app scope and endpoint filters.

**Comment by rudrankriyam** [comment]
@coderabbitai review

**Comment by rudrankriyam** [comment]
@codex review

File: internal/cli/optimize/keywords_score.go [hunk 7]
```diff
@@ -0,0 +1,533 @@
...
+Each keyword is evaluated against three sources, and any source that is
+unavailable is reported as unavailable instead of being replaced with an
+invented value:
+
+  Competition  Public App Store search plus public app metadata. No
+               authentication is required, and this source alone is enough to
+               produce a difficulty score.
+  Popularity   Official Apple Ads country-and-genre search demand. It requires
+               --genre plus Apple Ads credentials through --ad-account or
+               --ads-profile. It reads the most recent published 30-day window.
```
[the remark this answers is not in the document]
Comment (rudrankriyam) [reply internal/cli/optimize/keywords_score.go]: Fixed in b56c64809. Command help now says the source reads the latest complete week, and a focused help-contract test rejects the former 30-day wording.

File: internal/cli/optimize/keywords_difficulty.go [hunk 8]
```diff
@@ -0,0 +1,368 @@
...
+	phrase := normalizeKeywordText(keyword)
+	if phrase == "" {
+		return keywordMatchNone
+	}
+	normalizedTitle := normalizeKeywordText(title)
+	normalizedSubtitle := normalizeKeywordText(subtitle)
+	phraseTokens := strings.Fields(phrase)
+
+	switch {
+	case normalizedTitle != "" && strings.Contains(normalizedTitle, phrase):
```
[the remark this answers is not in the document]
Comment (rudrankriyam) [reply internal/cli/optimize/keywords_difficulty.go]: Fixed in b56c64809. Exact phrase checks now compare contiguous complete normalized tokens for title, subtitle, and combined text. Regression cases cover ai in Mail, app in Happy, and a cross-field substring.

File: internal/cli/optimize/keywords_score.go [hunk 9]
```diff
@@ -0,0 +1,533 @@
...
+
+	wanted := make(map[string]struct{}, len(request.Keywords))
+	for _, keyword := range request.Keywords {
+		wanted[keyword] = struct{}{}
+	}
+	popularity := make(map[string]asc.KeywordPopularity)
+	for _, row := range rows {
+		term := strings.ToLower(strings.Join(strings.Fields(row.Term), " "))
+		if _, ok := wanted[term]; !ok {
+			continue
```
[the remark this answers is not in the document]
Comment (rudrankriyam) [reply internal/cli/optimize/keywords_score.go:431]: Fixed in b56c64809. Requested terms and returned demand terms now use the same NFKC/token normalization while results remain keyed by the original requested keyword. Tests cover hyphenated input and full-width returned text.

File: internal/cli/optimize/keywords_score.go [hunk 10]
```diff
@@ -0,0 +1,533 @@
...
+		}
+		for appID, app := range result.Value {
+			metadata[appID] = app
+		}
+	}
+	return metadata, keywordPartialSourceStatus(
+		keywordSourceMetadata,
+		len(metadata),
+		failedChunks,
+		representativeKeywordError(failures),
```
[the remark this answers is not in the document]
Comment (rudrankriyam) [reply internal/cli/optimize/keywords_score.go]: Fixed in b56c64809. Successful lookup batches now compare every requested ID with returned metadata. Partial coverage remains available when evidence exists, while the source diagnostic reports the precise omitted count; the root command fixture asserts 4 of 5 IDs.

**Comment by rudrankriyam** [comment]
@coderabbitai review

**Comment by rudrankriyam** [comment]
@codex review

File: internal/asc/output_keyword_score.go [hunk 11]
```diff
@@ -0,0 +1,226 @@
...
+		rows = append(rows, []string{
+			row.Keyword,
+			formatKeywordScoreFloat(row.DifficultyScore),
+			formatKeywordScoreFloat(row.MinDifficultyScore),
+			formatKeywordScorePopularity(row.Popularity),
+			formatKeywordScoreInt(row.AppCount),
+			formatKeywordScoreInt(row.Rank),
+			formatKeywordScoreMatch(row.KeywordMatch),
+			formatKeywordScoreBool(row.IsBrandKeyword),
+			row.Status,
```
[the remark this answers is not in the document]
Comment (rudrankriyam) [reply internal/asc/output_keyword_score.go:123]: Fixed in 655f1add7. Registered table and Markdown keyword rows now include a compact per-row Error column, and the focused output test verifies a partial 429 diagnostic remains visible.

**Comment by rudrankriyam** [comment]
@codex review

**Comment by rudrankriyam** [comment]
@coderabbitai review

File: internal/cli/optimize/keywords_score.go [hunk 12]
```diff
@@ -0,0 +1,556 @@
...
+)
+
+var collectSearchPopularityForKeywords = ads.CollectSearchPopularity
+
+// KeywordsScoreCommand returns the composed keyword scoring command.
+func KeywordsScoreCommand() *ffcli.Command {
+	fs := flag.NewFlagSet("score", flag.ExitOnError)
+	keywords := shared.BindOnceCSVFlag(fs, "keywords", "Comma-separated keyword candidates to score (required)")
+	country := fs.String("country", "us", "ISO alpha-2 App Store storefront country or region")
+	appID := fs.String("app", "", "App Store app ID; adds this app's rank")
```
[the remark this answers is not in the document]
Comment (rudrankriyam) [reply internal/cli/optimize/keywords_score.go]: Fixed in 1504a3760. The --app flag description, long help rank behavior, and design source table now document ASC_APP_ID. A focused help-contract test pins both the long help and flag usage.

**Comment by rudrankriyam** [comment]
@codex review

**Comment by rudrankriyam** [comment]
@coderabbitai review

File: internal/cli/optimize/keywords_score.go [hunk 13]
```diff
@@ -0,0 +1,556 @@
...
+			popularity, popularitySource := collectKeywordPopularity(ctx, keywordPopularityRequest{
+				Keywords:   normalizedKeywords,
+				Country:    strings.ToUpper(normalizedCountry),
+				Genre:      normalizedGenre,
+				AdAccount:  *adAccount,
+				AdsProfile: *adsProfile,
+			})
+			sources = append(sources, popularitySource)
+
+			report := buildKeywordScoreReport(keywordScoreBuildInput{
```
[the remark this answers is not in the document]
Comment (rudrankriyam) [reply internal/cli/optimize/keywords_score.go:137]: Fixed in f34979dda: the command now returns ctx.Err() after collection and before report construction or output, with a partial-success cancellation regression.

File: internal/cli/optimize/keywords_score.go [hunk 14]
```diff
@@ -0,0 +1,556 @@
...
+				GeneratedAt: time.Now().UTC().Format(time.RFC3339),
+				AppID:       resolvedAppID,
+				Country:     strings.ToUpper(normalizedCountry),
+				Genre:       normalizedGenre,
+				Workers:     *workers,
```
[the remark this answers is not in the document]
Comment (rudrankriyam) [reply internal/cli/optimize/keywords_score.go]: Fixed in f34979dda: score now computes min(requested workers, normalized keyword count), uses it for both fan-outs, and reports the effective value.

File: internal/cli/optimize/keywords_score.go (hunk 13 above)
[the remark this answers is not in the document]
Comment (rudrankriyam) [reply internal/cli/optimize/keywords_score.go:137]: Fixed in 63bf250f7. Score now checks ctx.Err() after competition and popularity collection and before building or printing the report, returning cancellation without stdout. Added TestKeywordsScoreReturnsParentCancellationAfterPartialSuccess; focused and full local gates pass.

File: internal/cli/optimize/keywords_score.go (hunk 14 above)
[the remark this answers is not in the document]
Comment (rudrankriyam) [reply internal/cli/optimize/keywords_score.go]: Fixed in 63bf250f7. Score now computes effective workers as min(requested workers, normalized keyword count), uses that value for competition and metadata fan-out, and records it in the report. Added TestKeywordsScoreReportsEffectiveWorkerCount; focused and full local gates pass.

File: internal/cli/optimize/keywords_score.go [hunk 15]
```diff
@@ -0,0 +1,563 @@
...
+		metadataStatus,
+	}
+	if strings.TrimSpace(request.AppID) != "" {
+		ranked := 0
+		for _, entry := range competition {
+			if entry.Rank != nil {
+				ranked++
+			}
+		}
+		sources = append(sources, keywordSourceStatus(keywordSourceRank, ranked, nil))
```
[the remark this answers is not in the document]
Comment (rudrankriyam) [reply internal/cli/optimize/keywords_score.go]: Fixed in ca49fd5144e5528aa2919f7fc20810aa9a55a900. Partial app-rank coverage now uses the existing partial-source status with failed-search count and a representative error, so available results carry diagnostics and zero-rank partial failures are unavailable with an error. The score integration regression now asserts the 429 diagnostic; focused and full local gates pass.

File: internal/cli/optimize/keywords_score.go [hunk 16]
```diff
@@ -0,0 +1,563 @@
...
+Each keyword is evaluated against three sources, and any source that is
+unavailable is reported as unavailable instead of being replaced with an
+invented value:
+
+  Competition  Public App Store search plus public app metadata. No
+               authentication is required, and this source alone is enough to
+               produce a difficulty score.
+  Popularity   Official Apple Ads country-and-genre search demand. It requires
+               --genre plus Apple Ads credentials through --ad-account or
+               --ads-profile. It reads the latest complete week.
```
[the remark this answers is not in the document]
Comment (rudrankriyam) [reply internal/cli/optimize/keywords_score.go]: Fixed in ca49fd5144e5528aa2919f7fc20810aa9a55a900. Popularity help now separates Ads authentication (--ads-profile or ASC_ADS_PROFILE, access-token or key environment credentials, and the stored default profile) from ad-account selection (--ad-account, ASC_ADS_AD_ACCOUNT_ID, profile, or local account fallbacks). Added a help-contract regression; focused and full local gates pass.

File: internal/cli/optimize/keywords_score.go [hunk 17]
```diff
@@ -0,0 +1,576 @@
...
+	failedChunks := 0
+	omittedIDs := 0
+	for _, result := range results {
+		if result.Err != nil {
+			failedChunks++
+			failures = append(failures, result.Err)
+			continue
+		}
+		for appID, app := range result.Value {
+			metadata[appID] = app
```
[the remark this answers is not in the document]
Comment (rudrankriyam) [reply internal/cli/optimize/keywords_score.go:336]: Fixed in 6e22f3ded47ee6f38f0b9853a259c2601396d970. Public lookup hydration now accepts a competitor only when both releaseDate and currentVersionReleaseDate parse successfully; incomplete rows are excluded from the hydrated count, so the existing 365-day scoring fallback remains visible and the competitor_metadata source reports partial coverage with a required-release-metadata diagnostic. Added a regression covering missing and malformed dates, updated the existing partial-coverage contract, and reran focused plus full local gates.

File: internal/cli/optimize/keywords_metadata.go [hunk 18]
```diff
@@ -0,0 +1,140 @@
...
+		if result.TrackID == 0 {
+			continue
+		}
+		releaseDate := strings.TrimSpace(result.ReleaseDate)
+		currentVersionReleaseDate := strings.TrimSpace(result.CurrentVersionReleaseDate)
+		if _, ok := parsePublicDate(releaseDate); !ok {
+			continue
+		}
+		if _, ok := parsePublicDate(currentVersionReleaseDate); !ok {
+			continue
```
[the remark this answers is not in the document]
Comment (rudrankriyam) [reply internal/cli/optimize/keywords_metadata.go]: Fixed in 9966a465. Lookup hydration now retains each individually valid release date while blanking invalid fields; source coverage counts only rows with both valid dates and reports incomplete or missing date coverage. Added regressions proving partial dates still affect scoring/raw signals and reran focused plus full gates.

File: internal/asc/output_keyword_score.go [hunk 19]
```diff
@@ -0,0 +1,227 @@
...
+			compactKeywordScoreDiagnostic(row.Error),
+		})
+	}
+	render(
+		[]string{"Keyword", "Difficulty", "Min Difficulty", "Popularity", "Apps", "Rank", "Match", "Brand", "Status", "Error"},
+		rows,
```
[the remark this answers is not in the document]
Comment (rudrankriyam) [reply internal/asc/output_keyword_score.go:129]: Fixed in 9966a4656: the shared score table/Markdown renderer now includes an explicit Fallback column, with JSON/table/Markdown contract coverage.

[the remark this answers is not in the document]
Comment (rudrankriyam) [reply internal/asc/output_keyword_score.go:129]: Fixed in 9966a465. Keyword score table and Markdown output now include an explicit Fallback column, while JSON continues to preserve fallback:true. Added a JSON/table/Markdown renderer contract regression and reran focused plus full gates.
